reco_utils/dataset/movielens.py

```diff
@@ -377,48 +424,60 @@ def _get_schema(header, schema):
     return schema
 
 
-def _load_datafile(size, local_cache_path):
-    """ Download and extract file """
+def _maybe_download_and_extract(size, local_cache_path):
+    """Download and extract MovieLens rating and item datafiles if they don’t already exist.
 
+    Args:
+        size (str): Size of the data to load. One of ("100k", "1m", "10m", "20m").
+        local_cache_path (str): Path where to cache the zip file locally.
+            If None, use default, which is "data/ml-{size}/ml.zip"
+
+    Returns:
+        str: rating datafile path,
+        str: item datafile path.
+    """
     if size not in DATA_FORMAT:
         raise ValueError(ERROR_MOVIE_LENS_SIZE)
-    if not local_cache_path.endswith(".zip"):
+    if local_cache_path is None:
+        default_dir = os.path.join("data", "ml-{}".format(size))
+        local_cache_path = os.path.join(default_dir, "ml.zip")
+    elif not local_cache_path.endswith(".zip"):
         raise ValueError(ERROR_LOCAL_CACHE_PATH)
 
     path, filename = os.path.split(os.path.realpath(local_cache_path))
```

Comment:
I would suggest simplifying this section by leveraging tempfile.TemporaryDirectory and registering it's cleanup at process close. Also, I think it will be easier to cache the zip file only, decompression if fast, and that way you only have to keep track of a single file using local_cache_path.

```
from tempfile import TemporaryDirectory

tmp_dir = TemporaryDirectory()
tmp_path = tmp_dir.name

if local_cache_path is None:
  path = tmp_dir.name
  filename = "data.zip"
  atexit.register(tmp_dir.cleanup)
  ... download file to tmp dir...

... extract file to tmp dir ...
... load files from tmp dir ...
```

Reply:
Love this idea! cool! and yeah, totally agree that shutil.rmtree is very scary

Reply:
Checked my codes and recalled that we want to keep the downloaded data file (ratings and item data files) for later use. We don't need to keep the original zip file but `url_utils.maybe_download` uses urllib's `urlretrieve` which writes the downloaded data directly into the given file path. Can use w/ tempfile like [this](https://docs.python.org/3/howto/urllib2.html) which requires changes in `maybe_download` to use `shutil.copyfileobj()`

So I would like to keep the current mechanism for now. Instead, I'm going to refactor smoke and integration tests' shutil.rmtree to use tempdir you suggested, where we don't want to keep any downloaded data. What do you think @gramhagen?

Reply:
it's fine, i was just proposing to keep the zip file instead of the extracted files for simplicity, but it's one less thing to do if we keep the data files.

if we want to continue using urlretrieve (which is legacy, but I'm fine with retaining), then you can just give it a path in the temp dir

```
with TemporaryDirectory() as tmp_dir:
  filepath = maybe_download(url=URL, filename='data.zip', work_directory=tmp_dir.name)
  ... extract data from filepath to cache location...
```

Reply:
this is good. I like the solution.